src/pipecat/services/gladia/stt.py

```diff
@@ -293,36 +315,149 @@ def _prepare_settings(self) -> Dict[str, Any]:
+
+        # Send audio if connected
+        if self._connection_active and self._websocket and not self._websocket.closed:
+            await self._send_audio(audio)
+
+        yield None
+
+    async def _connection_handler(self):
+        """Handle WebSocket connection with automatic reconnection."""
+        while self._should_reconnect:
+            try:
+                # Initialize session if needed
+                if not self._session_url:
+                    settings = self._prepare_settings()
+                    response = await self._setup_gladia(settings)
+                    self._session_url = response["url"]
+                    self._reconnection_attempts = 0
+
+                # Connect with automatic reconnection
+                async for websocket in websockets.connect(self._session_url):
+                    try:
+                        self._websocket = websocket
+                        self._connection_active = True
+                        logger.info("Connected to Gladia WebSocket")
+
+                        # Send buffered audio if any
+                        await self._send_buffered_audio()
+
+                        # Start tasks
+                        receive_task = asyncio.create_task(self._receive_task_handler())
+                        keepalive_task = asyncio.create_task(self._keepalive_task_handler())
+
+                        # Wait for tasks to complete
+                        await asyncio.gather(receive_task, keepalive_task)
+
+                    except websockets.exceptions.ConnectionClosed as e:
+                        logger.warning(f"WebSocket connection closed: {e}")
+                        self._connection_active = False
+
+                        # Clean up tasks
+                        if "receive_task" in locals():
+                            receive_task.cancel()
+                        if "keepalive_task" in locals():
+                            keepalive_task.cancel()
+
+                        # Check if we should reconnect
+                        if not self._should_reconnect:
+                            break
+
+                        # Implement exponential backoff
+                        self._reconnection_attempts += 1
+                        if self._reconnection_attempts > self._max_reconnection_attempts:
+                            logger.error(
+                                f"Max reconnection attempts ({self._max_reconnection_attempts}) reached"
+                            )
+                            self._should_reconnect = False
+                            break
+
+                        delay = self._reconnection_delay * (2 ** (self._reconnection_attempts - 1))
+                        logger.info(
+                            f"Reconnecting in {delay} seconds (attempt {self._reconnection_attempts}/{self._max_reconnection_attempts})"
+                        )
+                        await asyncio.sleep(delay)
```

Comment:
We are repeating the reconnection logic inside the exception handler. It might be better to extract it into a helper function.

Or, we could make GladiaSTTService extend WebsocketService and rely on that for the reconnection logic. =)